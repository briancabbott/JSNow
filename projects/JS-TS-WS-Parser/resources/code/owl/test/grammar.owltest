#using owl.v1

# This is the grammar for Owl itself.
# Compile with `owl -c grammar.owl -o 1-parse.h`.

grammar = (rule | comment-token)*
rule = identifier '=' body
body = expr | (expr ':' identifier)+ operators*
operators = '.operators' fixity operator+
fixity =
  'postfix' : postfix-op
  'prefix' : prefix-op
  'infix' assoc : infix-op
assoc =
  'flat' : flat-op
  'left' : left-op
  'right' : right-op
  'nonassoc' : nonassoc-op
operator = expr ':' identifier
expr =
  identifier ('\\' ':' identifier@exception)* ('@' identifier@rename)? : ident
  string : literal
  [ '(' expr ')' ] : parens
  [ '[' string@begin-token expr\:choice? string@end-token ']' ] : bracketed
 .operators postfix
  '*' : zero-or-more
  '+' : one-or-more
  '?' : optional
 .operators infix flat
  '' : concatenation
 .operators infix flat
  '|' : choice
comment-token = 'line-comment-token' string
line-comment-token '#'
---#using owl.v1

# This is the grammar for Owl itself.
# Compile with `owl -c grammar.owl -o 1-parse.h`.

grammar = (rule | comment-token)*
rule = identifier '=' body
body = expr | (expr ':' identifier)+ operators*
operators = '.operators' fixity operator+
fixity =
  'postfix' : postfix-op
  'prefix' : prefix-op
  'infix' assoc : infix-op
assoc =
  'flat' : flat-op
  'left' : left-op
  'right' : right-op
  'nonassoc' : nonassoc-op
operator = expr ':' identifier
expr =
  identifier ('\\' ':' identifier@exception)* ('@' identifier@rename)? : ident
  string : literal
  [ '(' expr ')' ] : parens
  [ '[' string@begin-token expr\:choice? string@end-token ']' ] : bracketed
 .operators postfix
  '*' : zero-or-more
  '+' : one-or-more
  '?' : optional
 .operators infix flat
  '' : concatenation
 .operators infix flat
  '|' : choice
comment-token = 'line-comment-token' string
line-comment-token '#'
