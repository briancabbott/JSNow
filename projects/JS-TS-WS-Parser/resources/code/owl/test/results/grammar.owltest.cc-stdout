grammar (0 - 899)
  rule (104 - 137)
    identifier - grammar (104 - 111)
    body (114 - 137)
      expr : ZERO_OR_MORE (114 - 137)
        expr@operand : PARENS (114 - 136)
          expr : CHOICE (115 - 135)
            expr@operand : IDENT (115 - 119)
              identifier - rule (115 - 119)
            expr@operand : IDENT (122 - 135)
              identifier - comment-token (122 - 135)
  rule (138 - 164)
    identifier - rule (138 - 142)
    body (145 - 164)
      expr : CONCATENATION (145 - 164)
        expr@operand : IDENT (145 - 155)
          identifier - identifier (145 - 155)
        expr@operand : LITERAL (156 - 159)
          string - = (156 - 159)
        expr@operand : IDENT (160 - 164)
          identifier - body (160 - 164)
  rule (165 - 212)
    identifier - body (165 - 169)
    body (172 - 212)
      expr : CHOICE (172 - 212)
        expr@operand : IDENT (172 - 176)
          identifier - expr (172 - 176)
        expr@operand : CONCATENATION (179 - 212)
          expr@operand : ONE_OR_MORE (179 - 201)
            expr@operand : PARENS (179 - 200)
              expr : CONCATENATION (180 - 199)
                expr@operand : IDENT (180 - 184)
                  identifier - expr (180 - 184)
                expr@operand : LITERAL (185 - 188)
                  string - : (185 - 188)
                expr@operand : IDENT (189 - 199)
                  identifier - identifier (189 - 199)
          expr@operand : ZERO_OR_MORE (202 - 212)
            expr@operand : IDENT (202 - 211)
              identifier - operators (202 - 211)
  rule (213 - 254)
    identifier - operators (213 - 222)
    body (225 - 254)
      expr : CONCATENATION (225 - 254)
        expr@operand : LITERAL (225 - 237)
          string - .operators (225 - 237)
        expr@operand : IDENT (238 - 244)
          identifier - fixity (238 - 244)
        expr@operand : ONE_OR_MORE (245 - 254)
          expr@operand : IDENT (245 - 253)
            identifier - operator (245 - 253)
  rule (255 - 338)
    identifier - fixity (255 - 261)
    body (266 - 338)
      expr : LITERAL (266 - 275)
        string - postfix (266 - 275)
      expr : LITERAL (291 - 299)
        string - prefix (291 - 299)
      expr : CONCATENATION (314 - 327)
        expr@operand : LITERAL (314 - 321)
          string - infix (314 - 321)
        expr@operand : IDENT (322 - 327)
          identifier - assoc (322 - 327)
      identifier - postfix-op (278 - 288)
      identifier - prefix-op (302 - 311)
      identifier - infix-op (330 - 338)
  rule (339 - 432)
    identifier - assoc (339 - 344)
    body (349 - 432)
      expr : LITERAL (349 - 355)
        string - flat (349 - 355)
      expr : LITERAL (368 - 374)
        string - left (368 - 374)
      expr : LITERAL (387 - 394)
        string - right (387 - 394)
      expr : LITERAL (408 - 418)
        string - nonassoc (408 - 418)
      identifier - flat-op (358 - 365)
      identifier - left-op (377 - 384)
      identifier - right-op (397 - 405)
      identifier - nonassoc-op (421 - 432)
  rule (433 - 463)
    identifier - operator (433 - 441)
    body (444 - 463)
      expr : CONCATENATION (444 - 463)
        expr@operand : IDENT (444 - 448)
          identifier - expr (444 - 448)
        expr@operand : LITERAL (449 - 452)
          string - : (449 - 452)
        expr@operand : IDENT (453 - 463)
          identifier - identifier (453 - 463)
  rule (464 - 832)
    identifier - expr (464 - 468)
    body (473 - 832)
      expr : CONCATENATION (473 - 541)
        expr@operand : IDENT (473 - 483)
          identifier - identifier (473 - 483)
        expr@operand : ZERO_OR_MORE (484 - 516)
          expr@operand : PARENS (484 - 515)
            expr : CONCATENATION (485 - 514)
              expr@operand : LITERAL (485 - 489)
                string - \ (485 - 489)
              expr@operand : LITERAL (490 - 493)
                string - : (490 - 493)
              expr@operand : IDENT (494 - 514)
                identifier - identifier (494 - 504)
                identifier@rename - exception (505 - 514)
        expr@operand : OPTIONAL (517 - 541)
          expr@operand : PARENS (517 - 540)
            expr : CONCATENATION (518 - 539)
              expr@operand : LITERAL (518 - 521)
                string - @ (518 - 521)
              expr@operand : IDENT (522 - 539)
                identifier - identifier (522 - 532)
                identifier@rename - rename (533 - 539)
      expr : IDENT (552 - 558)
        identifier - string (552 - 558)
      expr : BRACKETED (571 - 587)
        expr : IDENT (577 - 581)
          identifier - expr (577 - 581)
        string@begin_token - ( (573 - 576)
        string@end_token - ) (582 - 585)
      expr : BRACKETED (599 - 660)
        expr : CONCATENATION (605 - 654)
          expr@operand : IDENT (605 - 623)
            identifier - string (605 - 611)
            identifier@rename - begin-token (612 - 623)
          expr@operand : OPTIONAL (624 - 637)
            expr@operand : IDENT (624 - 636)
              identifier - expr (624 - 628)
              identifier@exception - choice (630 - 636)
          expr@operand : IDENT (638 - 654)
            identifier - string (638 - 644)
            identifier@rename - end-token (645 - 654)
        string@begin_token - [ (601 - 604)
        string@end_token - ] (655 - 658)
      identifier - ident (544 - 549)
      identifier - literal (561 - 568)
      identifier - parens (590 - 596)
      identifier - bracketed (663 - 672)
      operators (674 - 750)
        fixity : POSTFIX_OP (685 - 692)
        operator (695 - 713)
          expr : LITERAL (695 - 698)
            string - * (695 - 698)
          identifier - zero-or-more (701 - 713)
        operator (716 - 733)
          expr : LITERAL (716 - 719)
            string - + (716 - 719)
          identifier - one-or-more (722 - 733)
        operator (736 - 750)
          expr : LITERAL (736 - 739)
            string - ? (736 - 739)
          identifier - optional (742 - 750)
      operators (752 - 794)
        fixity : INFIX_OP (763 - 773)
          assoc : FLAT_OP (769 - 773)
        operator (776 - 794)
          expr : LITERAL (776 - 778)
            string -  (776 - 778)
          identifier - concatenation (781 - 794)
      operators (796 - 832)
        fixity : INFIX_OP (807 - 817)
          assoc : FLAT_OP (813 - 817)
        operator (820 - 832)
          expr : LITERAL (820 - 823)
            string - | (820 - 823)
          identifier - choice (826 - 832)
  rule (833 - 876)
    identifier - comment-token (833 - 846)
    body (849 - 876)
      expr : CONCATENATION (849 - 876)
        expr@operand : LITERAL (849 - 869)
          string - line-comment-token (849 - 869)
        expr@operand : IDENT (870 - 876)
          identifier - string (870 - 876)
  comment_token (877 - 899)
    string - # (896 - 899)
