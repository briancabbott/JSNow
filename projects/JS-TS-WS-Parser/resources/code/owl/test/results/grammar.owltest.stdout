. grammar = (rule       | comment-token)*
  |         |expr:ident   expr:ident---  
  |         |expr:choice---------------  
  |         expr:parens----------------- 
  |         expr:zero-or-more------------
  |         body-------------------------
  rule-----------------------------------

. rule = identifier '='          body      
  |      expr:ident expr:literal expr:ident
  |      expr:concatenation----------------
  |      body------------------------------
  rule-------------------------------------

. body = expr       | (expr       ':'          identifier)+ operators *       
  |      |            |expr:ident expr:literal expr:ident   |                 
  |      |            |expr:concatenation----------------   |                 
  |      |            expr:parens-------------------------  expr:ident        
  |      |            expr:one-or-more--------------------- expr:zero-or-more 
  |      expr:ident   expr:concatenation------------------------------------- 
  |      expr:choice--------------------------------------------------------- 
  |      body---------------------------------------------------------------- 
  rule----------------------------------------------------------------------- 

. operators = '.operators' fixity     operator  +      
  |           |            |          expr:ident       
  |           expr:literal expr:ident expr:one-or-more 
  |           expr:concatenation---------------------- 
  |           body------------------------------------ 
  rule------------------------------------------------ 

. fixity =
  rule----

.   'postfix'    : postfix-op
    expr:literal             
    body---------------------
  ---------------------------

.   'prefix'     : prefix-op
    expr:literal            
  --------------------------
  --------------------------

.   'infix'      assoc      : infix-op
    expr:literal expr:ident           
    expr:concatenation-----           
  ------------------------------------
  ------------------------------------

. assoc =
  rule---

.   'flat'       : flat-op
    expr:literal          
    body------------------
  ------------------------

.   'left'       : left-op
    expr:literal          
  ------------------------
  ------------------------

.   'right'      : right-op
    expr:literal           
  -------------------------
  -------------------------

.   'nonassoc'   : nonassoc-op
    expr:literal              
  ----------------------------
  ----------------------------

. operator = expr       ':'          identifier
  |          expr:ident expr:literal expr:ident
  |          expr:concatenation----------------
  |          body------------------------------
  rule-----------------------------------------

. expr =
  rule--

.   identifier ('\\'         ':'          identifier@exception)* (           ..
    |          |expr:literal expr:literal expr:ident----------   |             
    |          |expr:concatenation----------------------------   |             
    |          expr:parens-------------------------------------  expr:parens---
    expr:ident expr:zero-or-more-------------------------------- expr:optional-
    expr:concatenation---------------------------------------------------------
    body-----------------------------------------------------------------------
  -----------------------------------------------------------------------------
 .. '@'          identifier@rename)? : ident
    expr:literal expr:ident-------          
    expr:concatenation------------          
  ---------------------------------         
  ----------------------------------        
  ----------------------------------        
  ------------------------------------------
  ------------------------------------------

.   string     : literal
    expr:ident          
  ----------------------
  ----------------------

.   [ '(' expr       ')' ] : parens
    |     expr:ident               
    expr:bracketed--------         
  ---------------------------------
  ---------------------------------

.   [ '[' string@begin-token expr\:choice? string@end-token ']' ] : bracketed
    |     |                  expr:ident--  |                                 
    |     expr:ident-------- expr:optional expr:ident------                  
    |     expr:concatenation-------------------------------                  
    expr:bracketed-----------------------------------------------            
  ---------------------------------------------------------------------------
  ---------------------------------------------------------------------------

.  .operators postfix          
   |          fixity:postfix-op
   operators-------------------
  -----------------------------
  -----------------------------

.   '*'          : zero-or-more
    expr:literal               
    operator-------------------
  -----------------------------
  -----------------------------
  -----------------------------

.   '+'          : one-or-more
    expr:literal              
    operator------------------
  ----------------------------
  ----------------------------
  ----------------------------

.   '?'          : optional
    expr:literal           
    operator---------------
  -------------------------
  -------------------------
  -------------------------

.  .operators infix flat         
   |          |     assoc:flat-op
   |          fixity:infix-op----
   operators---------------------
  -------------------------------
  -------------------------------

.   ''           : concatenation
    expr:literal                
    operator--------------------
  ------------------------------
  ------------------------------
  ------------------------------

.  .operators infix flat         
   |          |     assoc:flat-op
   |          fixity:infix-op----
   operators---------------------
  -------------------------------
  -------------------------------

.   '|'          : choice
    expr:literal         
    operator-------------
  -----------------------
  -----------------------
  -----------------------

. comment-token = 'line-comment-token' string    
  |               expr:literal-------- expr:ident
  |               expr:concatenation-------------
  |               body---------------------------
  rule-------------------------------------------

. line-comment-token '#'
  comment-token---------
